ctapipe/io/hessio.py

```diff
@@ -91,178 +43,12 @@ def hessio_event_source(url, max_events=None, allowed_tels=None,
-                    # To handle ASTRI and dst files
-                    data.r0.tel[tel_id].adc_samples = \
-                        pyhessio.get_adc_sum(tel_id)[..., None]
-                data.r0.tel[tel_id].adc_sums = \
-                    pyhessio.get_adc_sum(tel_id)
-                data.mc.tel[tel_id].reference_pulse_shape = \
-                    pyhessio.get_ref_shapes(tel_id)
-
-                nsamples = pyhessio.get_event_num_samples(tel_id)
-                if nsamples <= 0:
-                    nsamples = 1
-                data.r0.tel[tel_id].num_samples = nsamples
-
-                # load the data per telescope/pixel
-                hessio_mc_npe = pyhessio.get_mc_number_photon_electron
-                data.mc.tel[tel_id].photo_electron_image \
-                    = hessio_mc_npe(telescope_id=tel_id)
-                data.mc.tel[tel_id].meta['refstep'] = \
-                    pyhessio.get_ref_step(tel_id)
-                data.mc.tel[tel_id].time_slice = \
-                    pyhessio.get_time_slice(tel_id)
-                data.mc.tel[tel_id].azimuth_raw = \
-                    pyhessio.get_azimuth_raw(tel_id)
-                data.mc.tel[tel_id].altitude_raw = \
-                    pyhessio.get_altitude_raw(tel_id)
-                data.mc.tel[tel_id].azimuth_cor = \
-                    pyhessio.get_azimuth_cor(tel_id)
-                data.mc.tel[tel_id].altitude_cor = \
-                    pyhessio.get_altitude_cor(tel_id)
-            yield data
-            counter += 1
-
-            if max_events and counter >= max_events:
-                pyhessio.close_file()
-                return
-
-
-def _build_subarray_info(data, pyhessio):
-    """
-    constructs a SubarrayDescription object from the info in an
-    EventIO/HESSSIO file
-
-    Parameters
-    ----------
-    data: DataContainer
-        data container to fill in
-
-    Returns
-    -------
-    SubarrayDescription :
-        instrumental information
     """
-    telescope_ids = list(pyhessio.get_telescope_ids())
-    subarray =  SubarrayDescription("MonteCarloArray")
-
-    for tel_id in telescope_ids:
-        try:
-
-            pix_pos = pyhessio.get_pixel_position(tel_id) * u.m
-            foclen = pyhessio.get_optical_foclen(tel_id) * u.m
-            mirror_area = pyhessio.get_mirror_area(tel_id) * u.m ** 2
-            num_tiles = pyhessio.get_mirror_number(tel_id)
-            tel_pos = pyhessio.get_telescope_position(tel_id) * u.m
 
-            tel = TelescopeDescription.guess(*pix_pos, foclen)
-            tel.optics.mirror_area = mirror_area
-            tel.optics.num_mirror_tiles = num_tiles
-            subarray.tels[tel_id] = tel
-            subarray.positions[tel_id] = tel_pos
+    reader = HessioFileReader(None, None,
+                              input_url=url, **params )
 
-        except HessioGeneralError:
-            pass
 
-    return subarray
\ No newline at end of file
+    return (x for x in reader)
```

Comment:
Why not return `reader`? It is also iterable and will emit `x` when iterated over.

Reply:
`next(source)` doesn't work in that case, which is the old behavior used in some scripts.  Eventually I want to just deprecate `hessio_event_source`, but it will take some time to update scripts. In the future, everything should use an EventFileReader

Reply:
Oh `next` does not work on a reader? I did not realize that. Can we repair that?

Reply:
As was discussed in #613, EventFileReader was changed to be an [iterable](https://docs.python.org/3/glossary.html#term-iterable), not an [iterator](https://docs.python.org/3/glossary.html#term-iterator) (which has the `__next__` method). 

If you want to add this functionality to EventFileReader I would suggest opening a PR so we can discuss the implementation. Currently with EventFileReader we can do the following:
```
for event in reader:
   if event.count == 3:
      break
for event in reader:
   if event.count == 2:
      break
``` 
Where the it starts from the beginning again instead of the last position, which I think is more intuitive (its like looping through a list). If we include a `__next__` method I would argue we do it in a way that does not break this functionality.